core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,225 @@
+
+    // create a fresh snapshot with changes from cherry pick snapshot and previous Table metadata
+    Map<String, String> summaryForNewSnapshot = new HashMap<>(summary(base));
+    // register this wap id as che
+    summaryForNewSnapshot.put(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+    Snapshot outputSnapshot = new BaseSnapshot(ops.io(),
+        outputSnapshotId, parentSnapshotId, System.currentTimeMillis(), cherryPickSnapshot.operation(),
+        summaryForNewSnapshot, ops.io().newInputFile(manifestList.location()));
+    TableMetadata updated = base.addStagedSnapshot(outputSnapshot);
+    ops.commit(base, updated);
+
+    return outputSnapshot;
+  }
+
+  private static String stagedWapId(Snapshot snapshot) {
+    return snapshot.summary() != null ? snapshot.summary().getOrDefault("wap.id", null) : null;
+  }
+
+  private OutputFile createManifestList(
+      Long parentSnapshotId,
+      long outputSnapshotId,
+      List<ManifestFile> manifestsForNewSnapshot) {
+    OutputFile manifestList = manifestListPath();
+    try (ManifestListWriter listWriter = new ManifestListWriter(
+        manifestList, outputSnapshotId, parentSnapshotId)) {
+
+      ManifestFile[] manifestFiles = new ManifestFile[manifestsForNewSnapshot.size()];
+      Tasks.range(manifestFiles.length)
+          .stopOnFailure().throwFailureWhenFinished()
+          .executeWith(ThreadPools.getWorkerPool())
+          .run(index ->
+              manifestFiles[index] = manifestsForNewSnapshot.get(index));
+
+      listWriter.addAll(Arrays.asList(manifestFiles));
+
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to write manifest list file");
+    }
+    return manifestList;
+  }
+
+  /**
+   * Apply the pending changes and commit.
+   * <p>
+   * Changes are committed by calling the underlying table's commit method.
+   * <p>
+   * Once the commit is successful, the updated table will be refreshed.
+   *
+   * @throws ValidationException   If the update cannot be applied to the current table metadata.
+   * @throws CommitFailedException If the update cannot be committed due to conflicts.
+   */
+  @Override
+  public void commit() {
+    // Todo: Need to add retry
+    Snapshot outputSnapshot = apply();
+    base = ops.refresh();
+    ops.commit(base, base.cherrypickFrom(outputSnapshot));
+  }
+
+  /**
+   * Looks for manifest entries that have append files from cherry-pick snapshot
+   * and creates a new ManifestFile with a new snapshot
+   */
+  private ManifestFile createManifestFromAppends(List<ManifestFile> inputManifests, long outputSnapshotId)
+      throws IOException {
+
+    OutputFile out = manifestPath(manifestCount.getAndIncrement());
+    // create a manifest writer with new snapshot id
+    ManifestWriter writer = new ManifestWriter(ops.current().spec(), out, outputSnapshotId);
+    try {
+      for (ManifestFile manifest : inputManifests) {
```

Comment:
This should filter the manifests that are scanned using metadata, like [`Snapshot.addedFiles`](https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/BaseSnapshot.java#L162-L167) does.